airflow/sensors/sql_sensor.py

```diff
@@ -63,5 +81,15 @@ def poke(self, context):
         self.log.info('Poking: %s (with parameters %s)', self.sql, self.parameters)
         records = hook.get_records(self.sql, self.parameters)
         if not records:
-            return False
-        return str(records[0][0]) not in ('0', '')
+            if self.fail_on_empty:
+                raise AirflowException("No rows returned, raising as per fail_on_empty flag")
+            else:
+                return False
+        first_cell = records[0][0]
+        if callable(self.failure):
+            if self.failure(first_cell):
+                raise AirflowException(
+                    "Failure criteria met. self.failure({}) returned True".format(first_cell))
```

Comment:
I think we should raise an exception if failure or success are not callable - ignoring it is not a good idea.